core/src/main/java/org/apache/calcite/rel/metadata/RelMdForeignKeys.java

```diff
@@ -0,0 +1,246 @@
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * RelMdReferentialConstraits supplies a default implementation of
+ * {@link RelMetadataQuery#getForeignKeys} for the standard logical algebra.
+ * The relNodes supported are the same to {@link RelMetadataQuery#getUniqueKeys(RelNode)}
+ */
+public class RelMdForeignKeys
+    implements MetadataHandler<BuiltInMetadata.ForeignKeys> {
+  public static final ImmutableBitSet EMPTY_BIT_SET = ImmutableBitSet.of();
+  public static final RelMetadataProvider SOURCE =
+      ReflectiveRelMetadataProvider.reflectiveSource(
+          new RelMdForeignKeys(), BuiltInMetadata.ForeignKeys.Handler.class);
+
+//~ Constructors -----------------------------------------------------------
+
+  private RelMdForeignKeys() {}
+
+//~ Methods ----------------------------------------------------------------
+
+  @Override public MetadataDef<BuiltInMetadata.ForeignKeys> getDef() {
+    return BuiltInMetadata.ForeignKeys.DEF;
+  }
+
+  public ImmutableBitSet getForeignKeys(Filter rel, RelMetadataQuery mq, boolean ignoreNulls) {
+    return mq.getForeignKeys(rel.getInput(), ignoreNulls);
+  }
+
+  public ImmutableBitSet getForeignKeys(Sort rel, RelMetadataQuery mq, boolean ignoreNulls) {
+    return mq.getForeignKeys(rel.getInput(), ignoreNulls);
+  }
+
+  public ImmutableBitSet getForeignKeys(Correlate rel, RelMetadataQuery mq,
+      boolean ignoreNulls) {
+    return mq.getForeignKeys(rel.getLeft(), ignoreNulls);
+  }
+
+  public ImmutableBitSet getForeignKeys(TableModify rel, RelMetadataQuery mq,
+      boolean ignoreNulls) {
+    return mq.getForeignKeys(rel.getInput(), ignoreNulls);
+  }
+
+  public ImmutableBitSet getForeignKeys(Join rel, RelMetadataQuery mq, boolean ignoreNulls) {
+    final RelNode left = rel.getLeft();
+    final RelNode right = rel.getRight();
+    if (!rel.getJoinType().projectsRight()) {
+      // only return the foreign keys from the LHS since a semijoin only
+      // returns the LHS
+      return mq.getForeignKeys(left, ignoreNulls);
+    }
+
+    int nLeftColumns = rel.getLeft().getRowType().getFieldList().size();
+    ImmutableBitSet outForeignKeys = ImmutableBitSet.of();
+
+    if (!rel.getJoinType().generatesNullsOnLeft() || ignoreNulls) {
+      ImmutableBitSet leftInputForeignKeys = mq.getForeignKeys(left, ignoreNulls);
+      outForeignKeys = outForeignKeys.union(leftInputForeignKeys);
+    }
+    if (!rel.getJoinType().generatesNullsOnRight() || ignoreNulls) {
+      ImmutableBitSet rightInputForeignKeys = mq.getForeignKeys(right, ignoreNulls);
+      ImmutableBitSet.Builder rightOutForeignKeys = ImmutableBitSet.builder();
+      for (int index : rightInputForeignKeys.asList()) {
+        rightOutForeignKeys.set(index + nLeftColumns);
+      }
+      outForeignKeys = outForeignKeys.union(rightOutForeignKeys.build());
+    }
+    return outForeignKeys;
+  }
+
+  public ImmutableBitSet getForeignKeys(Aggregate rel, RelMetadataQuery mq,
```

Comment:
What if in the aggregation we have something other than grouping sets?

Reply:
ForieignKeys info are available in a table which is often the bottom part fo a whole relNode.
The foreign key may be invalid when propagate derivation from the bottom to up.
I think that a foreign key propagate from the bottom to up in aggregate,
if it is in the groupset, it can maintain the validity，
if it is in the other position except groupset, it can't maintain the validity, such as count(), sum() aggregate function.